common4j/src/main/com/microsoft/identity/common/java/crypto/BasicDecryptor.java

```diff
@@ -60,26 +62,79 @@ public byte[] decrypt(@NonNull final Key key,
+                          final byte[] dataToBeDecrypted,
+                          final int tagLength,
+                          final byte[] aad) throws ClientException {
+        return performCryptoOperationAndUploadTelemetry(
+                CryptoObjectName.Cipher,
+                decryptAlgorithm,
+                mCryptoFactory,
+                new ICryptoOperation<byte[]>() {
+                    @Override
+                    public byte[] perform() throws ClientException {
+                        return decryptWithGcm(
+                                key, decryptAlgorithm, iv, dataToBeDecrypted, tagLength, aad
+                        );
+                    }
+                }
+        );
+    }
+
+    private byte[] decryptWithIv(@NonNull final Key key,
+                                 @NonNull final String decryptAlgorithm,
+                                 final byte[] iv,
+                                 byte[] dataToBeDecrypted)
             throws ClientException {
         final Cipher cipher = mCryptoFactory.getCipher(decryptAlgorithm);
-        try{
+        try {
             if (iv != null && iv.length > 0) {
                 final IvParameterSpec ivSpec = new IvParameterSpec(iv);
                 cipher.init(Cipher.DECRYPT_MODE, key, ivSpec);
             } else {
                 cipher.init(Cipher.DECRYPT_MODE, key);
             }
-                return cipher.doFinal(dataToBeDecrypted);
+            return cipher.doFinal(dataToBeDecrypted);
+        } catch (final BadPaddingException e) {
+            throw new ClientException(ClientException.BAD_PADDING, e.getMessage(), e);
+        } catch (final IllegalBlockSizeException e) {
+            throw new ClientException(ClientException.INVALID_BLOCK_SIZE, e.getMessage(), e);
+        } catch (final InvalidKeyException e) {
+            throw new ClientException(ClientException.INVALID_KEY, e.getMessage(), e);
+        } catch (final InvalidAlgorithmParameterException e) {
+            throw new ClientException(ClientException.INVALID_ALG_PARAMETER, e.getMessage(), e);
+        }
+    }
+
+    @SuppressWarnings("NewApi")
+    private byte[] decryptWithGcm(@NonNull final Key key,
+                                  @NonNull final String decryptAlgorithm,
+                                  final byte[] iv,
+                                  byte[] dataToBeDecrypted,
+                                  final int tagLength,
+                                  @Nullable final byte[] aad)
+            throws ClientException {
+        final Cipher cipher = mCryptoFactory.getCipher(decryptAlgorithm);
+        try {
+            final GCMParameterSpec spec = new GCMParameterSpec(tagLength * 8, iv);
+            cipher.init(Cipher.DECRYPT_MODE, key, spec);
+
+            if (aad != null) {
+                cipher.updateAAD(aad);
+            }
+
+            return cipher.doFinal(
+                    dataToBeDecrypted, iv.length, dataToBeDecrypted.length - iv.length
```

Comment:
>iv.length, dataToBeDecrypted.length - iv.length [](http://example.com/codeflow?start=39&length=47)

here also, do you expect iv to be part of dataToBeDecrypted?

Reply:
It should just be cipher.doFinal(dataToBeDecrypted) since you taking iv separately

Reply:
`iv` is being for making the parameter spec and not part of `cipher.doFinal` call. Here I'm just providing the length of the iv for offset.

If I make the change that you suggest then tests are failing. I have tests for these here: https://github.com/AzureAD/ad-accounts-for-android/pull/2147

Reply:
That's because in encrypt you are adding iv to final ciphertext you return.
            encryptedText.put(iv);
            encryptedText.put(cipherText);
I don't think we should do that.

Reply:
Why shouldn't we do that? I looked at two solutions online that seemed to be doing this and I followed the same algorithm.

Reply:
Synced offline. Addressed here: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1953/commits/bb1949549d262200094b11ee93c56ad7efec295e